File: 0065-valid-number/0065-valid-number.py
class Solution:
    def isNumber(self, s: str) -> bool:
        state = 0
        finals = {2, 3, 7, 8}
        transfer = {
            (0, 'blank'): 0, (0, 'sign'): 1, (0, 'digit'): 2, (0, '.'): 4,
            (1, 'digit'): 2, (1, '.'): 4,
            (2, 'digit'): 2, (2, '.'): 3, (2, 'e'): 5, (2, 'blank'): 8,
            (3, 'digit'): 3, (3, 'e'): 5, (3, 'blank'): 8,
            (4, 'digit'): 3,
            (5, 'sign'): 6, (5, 'digit'): 7,
            (6, 'digit'): 7,
            (7, 'digit'): 7, (7, 'blank'): 8,
            (8, 'blank'): 8
        }

        for c in s:
            if '0' <= c <= '9':
                t = 'digit'
            elif c in "+-":
                t = 'sign'
            elif c in "eE":
                t = 'e'
            elif c in ". ":
                t = c
            else:
                t = 'invalid'
            
            if (state, t) not in transfer:
                return False
            state = transfer[(state, t)]
        
        return state in finals

File: 0065-valid-number/0065-valid-number.ts
function isNumber(s: string): boolean {
    let state = Array(10).fill(0);
    const finals = 0b101101000;
    const transfer = [
        [ 0, 1, 6, 2, -1],
        [-1,-1, 6, 2, -1],
        [-1,-1, 3, -1, -1],
        [ 8,-1, 3, -1, 4],
        [-1, 7, 5, -1, -1],
        [ 8,-1, 5, -1, -1],
        [ 8,-1, 6, 3, 4],
        [-1,-1, 5, -1, -1],
        [ 8,-1,-1, -1, -1]
    ];
    let idx = 0;
    for (let c of s) {
        idx = make(c);
        if (idx < 0) return false;
        state[0] = transfer[state[0]][idx];
        if (state[0] < 0) return false;
    }
    return (finals & (1 << state[0])) > 0;
}

function make(c: string): number {
    switch (c) {
        case ' ': return 0;
        case '+':
        case '-': return 1;
        case '.': return 3;
        case 'e':
        case 'E': return 4;
        default: return (c >= '0' && c <= '9') ? 2 : -1;
    }
}
